### PR TITLE
HPCC-19579 Add information about activities to the wu details meta service

### PR DESCRIPTION
**Pull request by shamser**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by shamser** [comment]
@ghalliday Please can you review.

File: esp/services/ws_workunits/ws_workunitsService.cpp [hunk 1]
```diff
@@ -4023,14 +4024,22 @@ bool CWsWorkunitsEx::onWUDetailsMeta(IEspContext &context, IEspWUDetailsMetaRequ
...
+        for (WuAttr attr=WuAttr::WaKind; attr<WuAttr::WaMax; ++((unsigned &)attr))
         {
-            const char * s = queryWuAttributeName((WuAttr)i);
-            if (s && *s)
-                attributes.append(s);
+            Owned<IEspWUDetailsMetaProperty> property = createWUDetailsMetaProperty("","");
+            const char * s = queryWuAttributeName(attr);
+            assertex(s && *s);
+            property->setName(s);
+            if (isListAttribute((WuAttr)attr))
```
Comment (ghalliday) [inline esp/services/ws_workunits/ws_workunitsService.cpp]: trivial: cast no longer required (and 2 lines down)

File: esp/services/ws_workunits/ws_workunitsService.cpp [hunk 2]
```diff
@@ -4023,14 +4024,22 @@ bool CWsWorkunitsEx::onWUDetailsMeta(IEspContext &context, IEspWUDetailsMetaRequ
         }
         resp.setStatistics(statistics);
 
-        StringArray attributes;
-        for (unsigned i=WuAttr::WaAll+1; i<WuAttr::WaMax; ++i)
+        IArrayOf<IEspWUDetailsMetaProperty> properties;
+        for (WuAttr attr=WuAttr::WaKind; attr<WuAttr::WaMax; ++((unsigned &)attr))
```
Comment (ghalliday) [inline esp/services/ws_workunits/ws_workunitsService.cpp]: An alternative to the cast on the ++ would be

    WuAttr & operator++(WuAttr & x) { x = (WuAttr)(x+1); return x; }

defined in wuattr.hpp.

**Comment by ghalliday** [comment]
@GordonSmith does this provide all you need?
@shamser a few minor comments

File: esp/scm/ws_workunits.ecm [hunk 3]
```diff
@@ -2147,16 +2147,37 @@ ESPResponse [exceptions_inline] WUDetailsResponse
...
+
 ESPRequest WUDetailsMetaRequest
 {
 };
 
 ESPResponse [exceptions_inline] WUDetailsMetaResponse
 {
     ESParray<string, Statistic> Statistics;
-    ESParray<string, Attribute> Attributes;
+    ESParray<ESPStruct WUDetailsMetaProperty, Property> Properties;
```
[the remark this answers is not in the document]
Comment (shamser) [reply esp/scm/ws_workunits.ecm]: "Properties" now include Statistics and Attributes  (reflects the structure of WUDetails response)

**Comment by shamser** [comment]
@ghalliday

File: esp/scm/ws_workunits.ecm [hunk 4]
```diff
@@ -2148,15 +2148,37 @@ ESPResponse [exceptions_inline] WUDetailsResponse
...
+
+ESPStruct WUDetailsMetaProperty
+{
+    string Name;
+    ESPenum WUDetailsAttrValueType ValueType;
+};
+
+ESPStruct WUDetailsActivityInfo
+{
+    unsigned kind;
```
Comment (ghalliday) [inline esp/scm/ws_workunits.ecm]: "Kind" for case consistency

File: esp/services/ws_workunits/ws_workunitsService.cpp [hunk 5]
```diff
@@ -4014,39 +4015,66 @@ bool CWsWorkunitsEx::onWUDetailsMeta(IEspContext &context, IEspWUDetailsMetaRequ
...
                 measures.append(s);
         }
         resp.setMeasures(measures);
+        IArrayOf<IConstWUDetailsActivityInfo> activities;
+        for (unsigned kind=((unsigned)ThorActivityKind::TAKnone)+1; kind< TAKlast; ++kind)
+        {
+            Owned<IEspWUDetailsActivityInfo> activity = createWUDetailsActivityInfo("","");
+            const char * name = getActivityText(static_cast<ThorActivityKind>(kind));
+            assertex(name && *name);
+            activity->setKind((unsigned) kind);
```
Comment (ghalliday) [inline esp/services/ws_workunits/ws_workunitsService.cpp]: trivial: cast not required.

**Pull request merged**